File: backend/app/smart/oauth2.py
import base64
import hashlib
import secrets
from datetime import datetime, timedelta

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.fhir_store import AuthorizationCode, OAuthClient, OAuthToken


class SMARTAuthServer:
    def __init__(self):
        # In-memory launch context store: launch_token → {"patient_id": ..., "expires_at": ...}
        # Used for SMART EHR-Launch flow: the EHR generates a token tied to a patient,
        # the SMART app then redeems it via /authorize to get that patient's context.
        self._launch_contexts: dict[str, dict] = {}

    # ── Launch context (EHR-Launch flow) ──────────────────────────────────

    def create_launch_context(self, patient_id: str) -> str:
        token = secrets.token_urlsafe(24)
        self._launch_contexts[token] = {
            "patient_id": patient_id,
            "expires_at": datetime.utcnow() + timedelta(minutes=10),
        }
        return token

    def resolve_launch_context(self, token: str | None) -> str | None:
        """Look up the patient_id bound to a launch token. Returns None if missing/expired."""
        if not token:
            return None
        ctx = self._launch_contexts.get(token)
        if not ctx:
            return None
        if ctx["expires_at"] < datetime.utcnow():
            self._launch_contexts.pop(token, None)
            return None
        return ctx["patient_id"]

    # ── Client management ────────────────────────────────────────────────

    async def get_client(self, db: AsyncSession, client_id: str) -> OAuthClient | None:
        return (
            await db.execute(select(OAuthClient).where(OAuthClient.client_id == client_id))
        ).scalar_one_or_none()

    async def seed_demo_client(self, db: AsyncSession):
        """Ensure the built-in SMART clients exist and their redirect_uri
        whitelists are up to date. Idempotent: on every backend start we
        upsert the canonical URI list, so when the medical-note dev port
        moves (3000 → 3001 etc.) existing installs pick up the new URI
        without a manual DB migration."""
        # Demo client (for testing with our own frontend) — keep
        # localhost:3000 AND :3001 so either port works during transitions.
        demo_uris = [
            "http://localhost:3001/callback",
            "http://localhost:3001/launch",
            "http://localhost:3000/callback",
            "http://localhost:3000/launch",
        ]
        demo = await self.get_client(db, "demo-smart-app")
        if not demo:
            db.add(
                OAuthClient(
                    client_id="demo-smart-app",
                    client_name="Demo SMART App",
                    redirect_uris=demo_uris,
                    allowed_scopes=[
                        "openid",
                        "fhirUser",
                        "launch/patient",
                        "patient/*.read",
                        "offline_access",
                    ],
                    is_confidential=False,
                )
            )
        else:
            demo.redirect_uris = demo_uris

        # voho0000/medical-note-smart-on-fhir — registered to consume our FHIR data
        mn_uris = [
            "https://voho0000.github.io/medical-note-smart-on-fhir/smart/callback",
            "http://localhost:3001/smart/callback",
            "http://localhost:3000/smart/callback",  # kept for backward compat
        ]
        mn = await self.get_client(db, "my_web_app")
        if not mn:
            db.add(
                OAuthClient(
                    client_id="my_web_app",
                    client_name="Medical Note SMART App",
                    redirect_uris=mn_uris,
                    allowed_scopes=[
                        "launch",
                        "openid",
                        "fhirUser",
                        "patient/*.read",
                        "online_access",
                        "offline_access",
                    ],
                    is_confidential=False,
                )
            )
        else:
            mn.redirect_uris = mn_uris

        await db.commit()

    # ── Authorization Code ───────────────────────────────────────────────

    async def create_auth_code(
        self,
        db: AsyncSession,
        client_id: str,
        redirect_uri: str,
        scopes: list[str],
        patient_id: str | None,
        code_challenge: str | None,
        code_challenge_method: str | None,
    ) -> str:
        code = secrets.token_urlsafe(32)
        db.add(
            AuthorizationCode(
                code=code,
                client_id=client_id,
                redirect_uri=redirect_uri,
                scopes=scopes,
                patient_id=patient_id,
                code_challenge=code_challenge,
                code_challenge_method=code_challenge_method,
                expires_at=datetime.utcnow() + timedelta(minutes=10),
            )
        )
        await db.commit()
        return code

    # ── Token exchange ───────────────────────────────────────────────────

    async def exchange_code(
        self,
        db: AsyncSession,
        code: str,
        client_id: str,
        redirect_uri: str,
        code_verifier: str | None,
    ) -> dict | None:
        row = (
            await db.execute(
                select(AuthorizationCode).where(
                    AuthorizationCode.code == code,
                    AuthorizationCode.used == False,  # noqa: E712
                )
            )
        ).scalar_one_or_none()

        if not row:
            return None
        if row.client_id != client_id or row.redirect_uri != redirect_uri:
            return None
        if row.expires_at < datetime.utcnow():
            return None

        # PKCE verification
        if row.code_challenge:
            if not code_verifier:
                return None
            if row.code_challenge_method == "S256":
                digest = hashlib.sha256(code_verifier.encode()).digest()
                computed = base64.urlsafe_b64encode(digest).rstrip(b"=").decode()
            else:
                computed = code_verifier
            if computed != row.code_challenge:
                return None

        row.used = True
        access_token = secrets.token_urlsafe(32)
        db.add(
            OAuthToken(
                client_id=client_id,
                access_token=access_token,
                scopes=row.scopes,
                patient_id=row.patient_id,
                expires_at=datetime.utcnow() + timedelta(hours=1),
            )
        )
        await db.commit()

        response: dict = {
            "access_token": access_token,
            "token_type": "Bearer",
            "expires_in": 3600,
            "scope": " ".join(row.scopes),
        }
        if row.patient_id:
            response["patient"] = row.patient_id
        return response

    # ── Token validation ─────────────────────────────────────────────────

    async def validate_token(self, db: AsyncSession, token: str) -> OAuthToken | None:
        row = (
            await db.execute(select(OAuthToken).where(OAuthToken.access_token == token))
        ).scalar_one_or_none()
        if row and row.expires_at > datetime.utcnow():
            return row
        return None


smart_auth = SMARTAuthServer()

File: backend-ts/src/smart/oauth2.ts
/**
 * SMART on FHIR OAuth2 server.
 *
 * Port of `backend/app/smart/oauth2.py`. Supports:
 *   - Launch context: pre-bind a patient ID to a short-lived token so
 *     the EHR-Launch flow targets the right patient.
 *   - Client management: lookup + idempotent seed of demo clients.
 *   - Authorization code flow: create + exchange (with PKCE verification).
 *   - Bearer token validation.
 */

import { createHash, randomBytes, timingSafeEqual } from "node:crypto";

import { and, eq } from "drizzle-orm";

import { type DB, db as defaultDb } from "@/core/database";
import {
  type OAuthClient,
  type OAuthToken,
  authorizationCodes,
  oauthClients,
  oauthTokens,
} from "@/models/schema";

function tokenUrlsafe(bytes: number): string {
  // Mirror Python's secrets.token_urlsafe — base64url, no padding.
  return randomBytes(bytes).toString("base64url");
}

/**
 * Single source of truth for the SMART discovery document. Served
 * verbatim from both `/fhir/.well-known/smart-configuration` and
 * `/smart/.well-known/smart-configuration` — these two paths are
 * historical (SMART App Launch IG §3.1 puts it at the FHIR base;
 * some clients also probe the authorization base) and must return
 * identical JSON so SMART apps see the same capability set
 * regardless of which path they hit.
 *
 * Only advertise what is actually implemented (audit 2026-06-12 P3):
 * no id_token is ever issued (so no sso-openid-connect capability and
 * no openid/fhirUser scopes) and no refresh_token is ever issued (so
 * no online_access/offline_access scopes). Advertising them makes
 * conforming clients mis-negotiate.
 */
export function buildSmartConfiguration(baseUrl: string): Record<string, unknown> {
  return {
    issuer: baseUrl,
    authorization_endpoint: `${baseUrl}/smart/authorize`,
    token_endpoint: `${baseUrl}/smart/token`,
    capabilities: [
      "launch-standalone",
      "launch-ehr",
      "client-public",
      "context-standalone-patient",
      "permission-patient",
    ],
    scopes_supported: ["launch", "launch/patient", "patient/*.read"],
    response_types_supported: ["code"],
    code_challenge_methods_supported: ["S256"],
  };
}

function nowPlusMinutes(min: number): Date {
  return new Date(Date.now() + min * 60_000);
}

function nowPlusHours(hr: number): Date {
  return new Date(Date.now() + hr * 60 * 60_000);
}

export class SMARTAuthServer {
  /** Launch contexts kept in memory (per Python parity — short TTL, no DB). */
  private launchContexts: Map<string, { patientId: string; expiresAt: Date }> = new Map();

  // ── Launch context ─────────────────────────────────────────────────

  createLaunchContext(patientId: string): string {
    const token = tokenUrlsafe(24);
    this.launchContexts.set(token, {
      patientId,
      expiresAt: nowPlusMinutes(10),
    });
    return token;
  }

  resolveLaunchContext(token: string | null | undefined): string | null {
    if (!token) return null;
    const ctx = this.launchContexts.get(token);
    if (!ctx) return null;
    // Single-use (audit 2026-06-12 P3): consume the token on first
    // resolve so a leaked launch token can't be replayed for the rest
    // of its 10-minute TTL.
    this.launchContexts.delete(token);
    if (ctx.expiresAt < new Date()) return null;
    return ctx.patientId;
  }

  // ── Client management ──────────────────────────────────────────────

  getClient(clientId: string, dbi: DB = defaultDb): OAuthClient | null {
    return dbi.select().from(oauthClients).where(eq(oauthClients.clientId, clientId)).get() ?? null;
  }

  /**
   * Ensure the built-in SMART clients exist and their redirect_uri
   * whitelists are up to date. Idempotent — re-running upserts the
   * canonical URI list.
   */
  seedDemoClients(dbi: DB = defaultDb): void {
    // Demo client (localhost dev frontend).
    // /launch entries pruned 2026-06-12 — they were redirect URIs
    // pointing at the (now removed) pre-PKCE frontend /launch page;
    // OAuth redirect_uris are callbacks, not launch entry points.
    const demoUris = ["http://localhost:3001/callback", "http://localhost:3000/callback"];
    const demoExisting = this.getClient("demo-smart-app", dbi);
    if (!demoExisting) {
      dbi
        .insert(oauthClients)
        .values({
          clientId: "demo-smart-app",
          clientName: "Demo SMART App",
          redirectUris: demoUris,
          allowedScopes: [
            "openid",
            "fhirUser",
            "launch/patient",
            "patient/*.read",
            "offline_access",
          ],
          isConfidential: false,
        })
        .run();
    } else {
      dbi
        .update(oauthClients)
        .set({ redirectUris: demoUris })
        .where(eq(oauthClients.clientId, "demo-smart-app"))
        .run();
    }

    // medical-note SMART app on GitHub Pages.
    const mnUris = [
      "https://voho0000.github.io/medical-note-smart-on-fhir/smart/callback",
      "http://localhost:3001/smart/callback",
      "http://localhost:3000/smart/callback",
    ];
    const mnExisting = this.getClient("my_web_app", dbi);
    if (!mnExisting) {
      dbi
        .insert(oauthClients)
        .values({
          clientId: "my_web_app",
          clientName: "Medical Note SMART App",
          redirectUris: mnUris,
          allowedScopes: [
            "launch",
            "openid",
            "fhirUser",
            "patient/*.read",
            "online_access",
            "offline_access",
          ],
          isConfidential: false,
        })
        .run();
    } else {
      dbi
        .update(oauthClients)
        .set({ redirectUris: mnUris })
        .where(eq(oauthClients.clientId, "my_web_app"))
        .run();
    }
  }

  // ── Authorization code ─────────────────────────────────────────────

  createAuthCode(
    {
      clientId,
      redirectUri,
      scopes,
      patientId,
      codeChallenge,
      codeChallengeMethod,
    }: {
      clientId: string;
      redirectUri: string;
      scopes: string[];
      patientId: string | null;
      codeChallenge: string | null;
      codeChallengeMethod: string | null;
    },
    dbi: DB = defaultDb,
  ): string {
    const code = tokenUrlsafe(32);
    dbi
      .insert(authorizationCodes)
      .values({
        code,
        clientId,
        redirectUri,
        scopes,
        patientId,
        codeChallenge,
        codeChallengeMethod,
        expiresAt: nowPlusMinutes(10),
        used: false,
      })
      .run();
    return code;
  }

  // ── Token exchange ─────────────────────────────────────────────────

  exchangeCode(
    {
      code,
      clientId,
      redirectUri,
      codeVerifier,
    }: {
      code: string;
      clientId: string;
      redirectUri: string;
      codeVerifier: string | null;
    },
    dbi: DB = defaultDb,
  ): Record<string, any> | null {
    const row = dbi
      .select()
      .from(authorizationCodes)
      .where(and(eq(authorizationCodes.code, code), eq(authorizationCodes.used, false)))
      .get();

    if (!row) return null;
    if (row.clientId !== clientId || row.redirectUri !== redirectUri) return null;
    if (row.expiresAt < new Date()) return null;

    // PKCE verification. timingSafeEqual over digests (audit 2026-06-12)
    // instead of `!==`, which leaks match-prefix length through timing.
    if (row.codeChallenge) {
      if (!codeVerifier) return null;
      let computed: string;
      if (row.codeChallengeMethod === "S256") {
        const digest = createHash("sha256").update(codeVerifier).digest();
        computed = digest.toString("base64url");
      } else {
        computed = codeVerifier;
      }
      const ca = createHash("sha256").update(computed).digest();
      const cb = createHash("sha256").update(row.codeChallenge).digest();
      if (!timingSafeEqual(ca, cb)) return null;
    }

    dbi
      .update(authorizationCodes)
      .set({ used: true })
      .where(eq(authorizationCodes.code, code))
      .run();

    const accessToken = tokenUrlsafe(32);
    dbi
      .insert(oauthTokens)
      .values({
        clientId,
        accessToken,
        scopes: row.scopes,
        patientId: row.patientId,
        expiresAt: nowPlusHours(1),
      })
      .run();

    const response: Record<string, any> = {
      access_token: accessToken,
      token_type: "Bearer",
      expires_in: 3600,
      scope: (row.scopes ?? []).join(" "),
    };
    if (row.patientId) response.patient = row.patientId;
    return response;
  }

  // ── Token validation ───────────────────────────────────────────────

  validateToken(token: string, dbi: DB = defaultDb): OAuthToken | null {
    const row = dbi.select().from(oauthTokens).where(eq(oauthTokens.accessToken, token)).get();
    if (row && row.expiresAt > new Date()) return row;
    return null;
  }
}

export const smartAuth = new SMARTAuthServer();
